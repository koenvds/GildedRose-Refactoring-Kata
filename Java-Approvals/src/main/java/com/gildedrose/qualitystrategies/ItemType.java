package com.gildedrose.qualitystrategies;

public class ItemType {
	public static final String AGED_BRIE = "Aged Brie";
	public static final String BACK_STAGE_PASS = "Backstage passes to a TAFKAL80ETC concert";
	public static final String SULFURAS = "Sulfuras, Hand of Ragnaros";
	public static final String CONJURED = "Conjured Mana Cake";
}
