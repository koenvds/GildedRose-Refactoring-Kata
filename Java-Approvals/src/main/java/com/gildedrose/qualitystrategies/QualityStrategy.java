package com.gildedrose.qualitystrategies;

import com.gildedrose.Item;

public interface QualityStrategy {
	public void updateItem(Item item);
}
