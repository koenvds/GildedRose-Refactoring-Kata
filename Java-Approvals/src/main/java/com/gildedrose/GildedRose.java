package com.gildedrose;

import com.gildedrose.qualitystrategies.QualityStrategyFactory;
import com.gildedrose.qualitystrategies.QualityStrategy;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (int i = 0; i < items.length; i++) {
        	Item item = items[i];
        	QualityStrategy strategy = QualityStrategyFactory.createQualityStrategy(item);
        	strategy.updateItem(item);
        }
    }
}