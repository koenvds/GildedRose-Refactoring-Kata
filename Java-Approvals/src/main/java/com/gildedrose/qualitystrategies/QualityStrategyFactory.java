package com.gildedrose.qualitystrategies;

import java.util.HashMap;
import java.util.Map;

import com.gildedrose.Item;
import com.gildedrose.qualitystrategies.impl.AgedItemQualityStrategy;
import com.gildedrose.qualitystrategies.impl.BackStagePassQualityStrategy;
import com.gildedrose.qualitystrategies.impl.ConjuredItemQualityStrategy;
import com.gildedrose.qualitystrategies.impl.NormalItemQualityStrategy;
import com.gildedrose.qualitystrategies.impl.LegendaryItemQualityStrategy;

public class QualityStrategyFactory {
	private static Map<String, QualityStrategy> qualityStrategies = new HashMap<String, QualityStrategy>();

	private QualityStrategyFactory() {
	}

	private static QualityStrategy defaultStrategy;
	static {
		defaultStrategy = new NormalItemQualityStrategy();

		qualityStrategies.put(ItemType.AGED_BRIE, new AgedItemQualityStrategy());
		qualityStrategies.put(ItemType.BACK_STAGE_PASS, new BackStagePassQualityStrategy());
		qualityStrategies.put(ItemType.CONJURED, new ConjuredItemQualityStrategy());
		qualityStrategies.put(ItemType.SULFURAS, new LegendaryItemQualityStrategy());
	}

	/**
	 * Returns the quality strategy depending on the item type received. Defaults to
	 * the NormalItemStrategy if no strategy is found.
	 * 
	 * @param item the item that a quality strategy needs to be found for.
	 * @return the quality strategy.
	 */
	public static QualityStrategy createQualityStrategy(Item item) {
		QualityStrategy strategy = qualityStrategies.get(item.name);
		if (strategy == null) {
			strategy = defaultStrategy;
		}
		return strategy;
	}

}
