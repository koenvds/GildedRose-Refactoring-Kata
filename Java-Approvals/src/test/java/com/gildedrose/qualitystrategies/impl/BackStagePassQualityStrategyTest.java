package com.gildedrose.qualitystrategies.impl;


import java.util.ArrayList;
import java.util.List;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.gildedrose.Item;
import com.gildedrose.qualitystrategies.ItemType;

@UseReporter(DiffReporter.class)
public class BackStagePassQualityStrategyTest {
	private static BackStagePassQualityStrategy qualityStrategy;

	private static final String itemName = ItemType.BACK_STAGE_PASS;

	@BeforeAll
	private static void initialize() {
		qualityStrategy = new BackStagePassQualityStrategy();
	}

	@Test
	public void testBackstagePassesQualityIncreasesBy1Upto11DaysToSellinDate() {
		//Arrange
		List<Item> backstagePasses = new ArrayList<>();
		for(int sellinDays = 12; sellinDays >= 11; sellinDays--) {
			backstagePasses.add(new Item(itemName, sellinDays, 0));
		}

		//Act
		backstagePasses.stream().forEach(qualityStrategy::updateItem);

		//Assert
		Approvals.verifyAll("Item sellin going from 12 to 11 ==> quality up by 1 (from 0)", "items", backstagePasses);
	}

	@Test
	public void testBackstagePassesQualityIncreasesBy2Between10And6DaysToSellinDate() {
		//Arrange
		List<Item> backstagePasses = new ArrayList<>();
		for(int sellinDays = 10; sellinDays >= 6; sellinDays--) {
			backstagePasses.add(new Item(itemName, sellinDays, 0));
		}

		//Act
		backstagePasses.stream().forEach(qualityStrategy::updateItem);

		//Assert
		Approvals.verifyAll("Item sellin going from 10 to 6 ==> quality up by 2 (from 0)", "items", backstagePasses);
	}

	@Test
	public void testBackstagePassesQualityIncreasesBy3Between5And1DayToSellinDate() {
		//Arrange
		List<Item> backstagePasses = new ArrayList<>();
		for(int sellinDays = 5; sellinDays >= 1; sellinDays--) {
			backstagePasses.add(new Item(itemName, sellinDays, 0));
		}

		//Act
		backstagePasses.stream().forEach(qualityStrategy::updateItem);

		//Assert
		Approvals.verifyAll("Item sellin going from 5 to 1 ==> quality up by 3 (from 0)", "items", backstagePasses);
	}


	@Test
	public void testBackstagePassesQualityGoesToZeroPassedSellinDate() {
		//Arrange
		var expected = """
			Quality drops to 0 (sellIn: 0 -> -1, quality: 10 -> 0)
					""";

		int startSellIn = 0;
		int startQuality = 10;
	
		Item item = new Item(itemName, startSellIn, startQuality);

		//Act
		qualityStrategy.updateItem(item);

		//Assert
		Approvals.verify(String.format("Quality drops to 0 (sellIn: %s -> %s, quality: %s -> %s)", startSellIn, item.sellIn, startQuality, item.quality), new Options().inline(expected));
	}

}
