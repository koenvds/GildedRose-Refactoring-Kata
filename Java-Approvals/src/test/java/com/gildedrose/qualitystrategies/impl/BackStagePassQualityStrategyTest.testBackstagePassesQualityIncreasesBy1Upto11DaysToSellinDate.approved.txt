Item sellin going from 12 to 11 ==> quality up by 1 (from 0)


items[0] = Backstage passes to a TAFKAL80ETC concert, 11, 1
items[1] = Backstage passes to a TAFKAL80ETC concert, 10, 1
